File: ternarysearch.java
/**
 ** Java Program to Implement Ternary Search Algorithm
 **/
 
import java.util.Scanner;
 

public class TernarySearch
{
  
    public static int ternarySearch (int[] A, int value)
    {
        return ternarySearch(A, value, 0, A.length - 1);
    }
  
    public static int ternarySearch (int[] A, int value, int start, int end)
    {
        if (start > end) 
            return -1;       
 
     
        int mid1 = start + (end-start) / 3;        
       
        int mid2 = start + 2*(end-start) / 3;
 
        if (A[mid1] == value) 
            return mid1;
 
        else if (A[mid2] == value) 
            return mid2;
     
        else if (value < A[mid1]) 
            return ternarySearch (A, value, start, mid1-1);
       
        else if (value > A[mid2]) 
            return ternarySearch (A, value, mid2+1, end);
      
        else 
            return ternarySearch (A, value, mid1,mid2);        
    }
 
    public static void main(String[] args) 
    {
        Scanner scan = new Scanner( System.in );        
        System.out.println("Ternary Search Test\n");
        int n, i;
       
        System.out.println("Enter number of integer elements");
        n = scan.nextInt();
       
        int arr[] = new int[ n ];
       
        System.out.println("\nEnter "+ n +" sorted integer elements");
        for (i = 0; i < n; i++)
            arr[i] = scan.nextInt();
        System.out.println("\nEnter element to search for : ");
        int key = scan.nextInt();
 
        int result = ternarySearch(arr, key);
 
        if (result == -1)
            System.out.println("\n"+ key +" element not found");
        else
            System.out.println("\n"+ key +" element found at position "+ result);
 
    }    
}

File: Ternary_Search.cpp
//SATYAMEVA JAYATE 		

			// Used to find max value in unimodal function

#include <bits/stdc++.h>
using namespace std;
int arr[100];
int ternary(int l,int r,int key)
{
	int mid1=(r+2*l)/3; // Using Section formula
	int mid2=(2*r+l)/3;
	if(l<=r)
	{
		if(arr[mid1]==key)
			return mid1;
		if(arr[mid2]==key)
			return mid2;
		else
		{
			if(arr[mid1]>key)
		   	   return ternary(l,mid1-1,key);
			else if(arr[mid2] <key)
				return ternary(mid2+1,r,key);
			else 
				return ternary(mid1+1,mid2-1,key);
		}
	}
	return -1;
}
int main()
{
	int i,n,key,ans;
	cin >> n; // Element in the array
	for(i=0;i<n;i++)
	{
		cin >> arr[i];
	}
	sort(arr,arr+n);// Hence the array is sorted
	cout <<"Enter the element to be searched\n";
	cin >> key;
	ans = ternary(0,n,key);
	cout <<ans<<"\n";
	return 0;
}

